teps/0028-task-results-without-results.md

```diff
@@ -0,0 +1,358 @@
+    - [Story 3](#story-3)
+    - [Story 4](#story-4)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Alternatives](#alternatives)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+A `task` in a `pipeline` can produce a result and that result can be consumed in many ways within that `pipeline`:
+
+* `param` mapping in a consumer `pipelineTask`
+
+```yaml
+kind: Pipeline
+spec:
+  tasks:
+    - name: format-result
+      taskRef:
+        name: format-result
+      params:
+        - name: result
+          value: "$(tasks.sum-inputs.results.result)"
+```
+
+* When Expression
+
+```yaml
+kind: Pipeline
+spec:
+  tasks:
+    - name: echo-file-exists
+      when:
+        - input: "$(tasks.check-file.results.exists)"
+          operator: in
+          values: ["yes"]
+```
+
+* Pipeline Results
+
+```yaml
+kind: Pipeline
+spec:
+  tasks:
+    ...
+  results:
+    - name: sum
+      value: $(tasks.second-add.results.sum)
+```
+
+Today, `pipeline` is declared `failure` and stops executing further after task result resolution fails because of
+missing task results. There are many reasons for a missing task result:
+ 
+* a `task` producing task result failed, no result available
+* a `task` producing result was skipped/disabled and no result generated
+* a `task` producing result did not generate that result even without any failure
+
+As we are expanding task results consumption within `finally` tasks and adding support for `continueAfterSkip`,
+it becomes crucial to define how a `pipeline` can handle such missing results. In this proposal, we are evaluating
+different possible solutions and determine which one fits best to address this issue.
+
+## Motivation
+
+Missing task results should not be fatal. Continue executing `pipeline` when a task result is missing.
+
+### Goals
+
+Identify a set of use cases for `finally` and `continueAfterSkip` without task results and design a common solution
+to address those use cases.
+
+### Non-Goals
+
+This proposal is not offering a solution without any use case.
+
+## Proposal
+
+Enable `pipeline` author to specify a default value for a `task result` in case of a missing result. Pipeline author
+can specify default for the `param` referring to task result if that `param` is declared optional in the `task`.
+
+Generally, task author designs a task with set of inputs in the form of `params` while producing some results in the
+form of `results`. When a `task` is incorporated in a `pipeline`, `pipeline` author has control over these `params`
+and can design it so that the `params` are mapped to `pipeline` params or initialized with the `results` of other
+tasks in a `pipeline`. `params` in a `task` can be declared as required and optional. Required `params` does not allow
+any defaults and have to be initialized at the `pipeline` or `pipelineRun` level.
+
+The proposal is designed on the similar lines, when a `param` is declared `optional` at the `task` level and referring to the
```

Comment:
I'm afraid I don't understand the reasoning behind this.
If a parameter is declared as optional in the `Task`, by definition it is acceptable not to pass any value to it.

Optional parameters are those least in need of a default, because if the expression injected by the pipeline depends on something not available (like a result), the pipeline could simply not pass any value, and it would work, because it's optional.

Mandatory parameters instead cause a Task failure if not passed. If the pipeline author binds a mandatory task parameter to a result from another task, we need a way for the pipeline author to supply a default value (if they want) in case the result in question is not available.

Reply:
> Optional parameters are those least in need of a default, because if the expression injected by the pipeline depends on something not available (like a result), the pipeline could simply not pass any value, and it would work, because it's optional.

Optional param at the task level referring to task result in a `Pipeline` is considered as `required` i.e. if a task result is not initiated, pipeline stops executing even if a referenced param was optional at the task level.